Add type info interfaces motivated by numba

aaronj0 opened the pull request:
- Add `IsSameType` used for matching arg types between numba inferred signatures and CppOverloads in cppyy.
- Add type info reflection interfaces for integer, void pointer, signed and unsigned types.
- Also added interfaces `IsIntegral` and `IsFloating` that check if types have the respective representations. Used in the case of Numba where scoring based on reflection is required.

will add tests

vgvassilev commented:
We should add tests.

jeaye commented:
Would love to have this. What's needed before we can merge?

aaronj0 commented:
> Would love to have this. What's needed before we can merge?

Hi! I have some unittests that I haven't pushed yet, and this PR should be ready to go. Will do in the next few hours once I have access to a laptop

vgvassilev commented on include/clang/Interpreter/CppInterOp.h:
```diff
@@ -509,15 +509,40 @@ namespace Cpp {
   CPPINTEROP_API bool IsPODType(TCppType_t type);
 
+  /// Checks if type has an integer representation
+  CPPINTEROP_API bool IsIntegerType(TCppType_t type);
```
Probably we should have an enum instead of many interfaces..

aaronj0 committed:
Add various type check interfaces required by numba extension

aaronj0 committed:
Add tests for new type check interfaces

Adds tests for IsSameType, IsIntegerType, IsFloatingType, GetUnqualifiedType, IsVoidPointerType

aaronj0 commented on include/CppInterOp/CppInterOp.h:
```diff
@@ -578,6 +579,22 @@ CPPINTEROP_API bool IsRecordType(TCppType_t type);
 
+/// Checks if type has an integer representation
+CPPINTEROP_API bool IsIntegerType(TCppType_t type,
+                                  Signedness s = Signedness::kAny);
```
Yes, what you describe will happen if the caller specifies signed or unsigned. I was of the opinion that if you call this by default (for kAny) it will return true if the qualtype has an integer representation which should cover the above scenario

aaronj0 replied:
Done, I've opened a fresh PR instead since this was hard to rebase: https://github.com/compiler-research/CppInterOp/pull/913

aaronj0 commented:
Closing in favour of https://github.com/compiler-research/CppInterOp/pull/913
